### PR TITLE
Core algorithm refactoring

### PR DESCRIPTION
**Pull request by JanWittler**
This PR aims to refactor the core algorithm to make it more readable and maintainable.
The following refactorings were performed:
- move state `Token.marked` into algorithm
- remove optimisations to reduce RAM usage in form of reversed hash tables only being generated when necessary
- remove `TokenHashMap` which only differed from a normal map by using a modulo on the keys which had no benefit for the algorithm
- encapsulate `Token.hash`, `TokenList.tokenHashes`, `TokenList.hashLength` into separate class `SubsequenceHashLookupTable` with improved naming and documentation
- remove `TokenList` and replace it with `List<Token>`
- move state `Token.isBaseCode` into algorithm
  - refactored base code tagging to occur outside of the core algorithm
  - only tag base code for compared submission, not for the base code submission
- refactors `JPlagComparison` to record

#### `TokenList` removal
With the removal of `TokenList` come two minor changes. 
First, the performed string optimisation to point to the same String object in case the file name for multiple tokens is the same was removed. I don't think this optimisation is necessary with modern-day computers anymore.
Second, a token list adjusted the tokens to have monotonically non-decreasing line numbers. This is only relevant for the visualisation, as the matching does not use the line number information. With some minor testing, the adjustment behaviour seemed to trigger only for FILE_END tokens which was fixed by assigning a NO_VALUE line to any FILE_END token. Since FILE_END token cannot appear in any match anyways, it should not make any difference for the visualisation of results.

#### Choice of actual `Set` / `Map` class
The refactorings introduce various variables of type `Set` or `Map`. To tune performance, we could test different types of these classes. An initial discussion was already started [here](https://github.com/jplag/JPlag/pull/534#discussion_r931576763). However, I would postpone this tuning until all changes (including follow-up PRs) are integrated.

#### Choice of subsequence hash constants

In `SubsequenceHashLookupTable`, we have the `MAX_HASH_LENGTH` and the `TOKEN_HASH_MODULO` parameters which are needed to restrict the computed hash to not exceed `Integer.MAX_VALUE`. Currently, the length is 25 and the modulo is 64. This means, that if a minimum match length of >25 is passed, the hashes are still only computed for 25, which may result in too many subsequence suggestions. The modulo implies that if a token type with a value of >64 is encountered, it is mapped to `value % 64`, which results in hash collision with another token type and thus may result in too many suggestions too. As the subsequences are only suggestions and verified afterwards, this is not a problem for functionality but for performance. Thus, we should re-evaluate whether a smaller `MAX_HASH_LENGTH` in favour of a larger `TOKEN_HASH_VALUE` is to be preferred (as we currently don't have anywhere a minimum token match of over 20). Exemplary, we could take the value pair `23 / 257`. We could also include multiple value pairs and select the appropriate one based on the passed `windowSize`. However, in local tests with the available example submissions, I could not identify a large benefit which is why I didn't change the values. We should test this with a large submission set. Also, this will be influenced by #623 as this PR will greatly decrease the times a token has a value larger than the MODULO. 

### Testing

As the test coverage of the core is not too high, we should validate the refactorings on additional data sets (and maybe try to integrate some of those in form of plain token sequences into the core). The refactorings should not change functionality, thus the similarity values should remain identical for any submission.
Furthermore, as the algorithm is highly optimised for performance, we should compare the performance of the refactored implementation to the previous one with a large data set.
Unfortunately, I don't have a comparison set at hand, so it would be beneficial if some of the @jplag/maintainer could help out here.

#### Tokens
I refactored all state related information of tokens into the algorithm. In a future PR, we should refactor tokens to be records instead of classes. However, for this to be possible, we need to avoid subclassing tokens for every language. This will probably not be needed anymore as soon as token constants are refactored to enums.

#### Naming
We use `firstSubmission` and `secondSubmission` throughout the application. From my perspective, these names indicate that there might be some kind of ordering which is not. Thus, I renamed the submissions in the core algorithm to `leftSubmission` and `rightSubmission`. Another option would have been `patternSubmission` and `textSubmission` which would be closer to the original paper. I chose the former naming as the clear separation between pattern (baseline file) and text (file to detect matches in) is not given in our application context. If left / right naming is accepted, in a following PR we should adjust the naming throughout the application.

#### Parallel Comparison Strategy
Since tokens are now state-free, it should be possible to entirely remove the locking the parallel comparison strategy. Will be done in #624.

### CODE REVIEW

**Comment by JanWittler** [comment]
https://github.com/jplag/JPlag/blob/0780fc49c5d37c70f172f9d8289ddc158b6de934/jplag.frontend-utils/src/main/java/de/jplag/TokenList.java#L34-L45

I would like to remove the `TokenList` data structure entirely for which we have to get rid of this custom add functionality.
The first part (ll37-39) with the file name can probably be removed.
For the second part (ll40-42) I am not sure whether we require a token sequence to have monotonically non-decreasing lines. I would guess if anywhere then in the report viewer. @nestabentum do you by accident know more about that?

**Comment by nestabentum** [comment]
@JanWittler all I know is that the `TokenList` is currently used to determine the matches of a comparison: https://github.com/jplag/JPlag/blob/0780fc49c5d37c70f172f9d8289ddc158b6de934/jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java#L156-L171

**Comment by JanWittler** [comment]
> @JanWittler all I know is that the `TokenList` is currently used to determine the matches of a comparison:
> 
> https://github.com/jplag/JPlag/blob/0780fc49c5d37c70f172f9d8289ddc158b6de934/jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java#L156-L171
> 
> The information on matches is then further used to highlight them in the comparison view.

Alright, thanks for the pointer. I will have a look at that.

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/GreedyStringTiling.java:181]: Consider an identity HashSet (otherwise we rely on equals and hashcode of the token, not sure whether we want this.

`Collections.newSetFromMap(new IdentityHashMap<>())` .. but not sure here

File: jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java:157]: maybe we want that getTokenList() returns immutable lists or copies .. just to make the impl clean.

**Review by dfuchss — COMMENTED** [review]
Just some minor comments :)

File: jplag.frontend.cpp/src/main/java/de/jplag/cpp/Scanner.java
Comment (JanWittler) [reply jplag.frontend.cpp/src/main/java/de/jplag/cpp/Scanner.java:13]: Not possible due to naming conflicts. There is another `Token` class in the package (I think generated from the grammar).

File: jplag.frontend.python-3/src/main/java/de/jplag/python3/Parser.java
Comment (JanWittler) [reply jplag.frontend.python-3/src/main/java/de/jplag/python3/Parser.java:25]: Not possible due to naming conflicts. There is another Token class in the package (I think generated from the grammar).

File: jplag.frontend.scheme/src/main/java/de/jplag/scheme/Parser.java
Comment (JanWittler) [reply jplag.frontend.scheme/src/main/java/de/jplag/scheme/Parser.java:13]: Not possible due to naming conflicts. There is another Token class in the package (I think generated from the grammar).

File: jplag.frontend.text/src/main/java/de/jplag/text/Parser.java
Comment (JanWittler) [reply jplag.frontend.text/src/main/java/de/jplag/text/Parser.java:25]: Not possible due to naming conflict with `antlr.Token`

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (JanWittler) [reply jplag/src/main/java/de/jplag/GreedyStringTiling.java:181]: I am aiming to refactor `Token` to a record class thus `equals` and `hashCode` should get generated correctly.
However, it might still makes sense to use an identity HashSet as reference equality suffices for our use case. Do you know how the performance between HashSet and IdentityHashSet compares?

Comment (dfuchss) [reply jplag/src/main/java/de/jplag/GreedyStringTiling.java:181]: HashSet uses equals, IdentityHashMap (so also the set) uses ==

File: jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java
Comment (JanWittler) [reply jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java:157]: Done.

**Comment by JanWittler** [comment]
The aim of this PR is to make the core algorithm understandable by looking at the code, without reading the paper. From my perspective this is now given, thus the PR is ready for review. If you spot any parts of the algorithm that you still don't understand or would like to see more documentation for, please comment.

**Comment by tsaglam** [comment]
> For the second part (ll40-42) I am not sure whether we require a token sequence to have monotonically non-decreasing lines. I would guess if anywhere then in the report viewer. @nestabentum do you by accident know more about that?

I think these lines ensure that EOF tokens have a proper line number. As EOF tokens are specified without a line index, this might be the here to set the line index to the one of the second to last token of the file. EOF tokens cannot easily given a line number in the frontend, as the line indexing is done by the parser/scanner/compiler. Also, this might be an optimization for tokens that do not use lines and columns but only a single index. However, this is less likely.

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (JanWittler) [reply jplag/src/main/java/de/jplag/GreedyStringTiling.java]: Will be overhauled in #623 anyways where we can't use identity anymore. Thus is it fine if we keep the `HashSet` here? We can later still do performance optimisation.

**Comment by JanWittler** [comment]
@jplag/maintainer 
I think there is no other PR anymore blocking this one, so it is ready to merge from my side. The end2end tests all run successfully, so I would assume that the changes do not change any results (as intended). However, if you like to validate this on a larger submission set, feel free to do so.

File: jplag.frontend-testutils/src/test/java/de/jplag/testutils/TokenUtils.java
Comment (dfuchss) [inline jplag.frontend-testutils/src/test/java/de/jplag/testutils/TokenUtils.java:15]:
```suggestion
     * @param tokenList is the list of {@link Token Tokens}.
```

File: jplag.frontend-testutils/src/test/java/de/jplag/testutils/TokenUtils.java
Comment (dfuchss) [inline jplag.frontend-testutils/src/test/java/de/jplag/testutils/TokenUtils.java:25]:
```suggestion
     * @param tokenList is the list of {@link Token Tokens}.
```

File: jplag.frontend-utils/src/main/java/de/jplag/Language.java
Comment (dfuchss) [inline jplag.frontend-utils/src/main/java/de/jplag/Language.java:37]: Maybe we should think about `List<? extends Token>` as this would make it clear that the list is only readable.

File: jplag.frontend.rust/src/test/java/de/jplag/rust/RustFrontendTest.java
Comment (dfuchss) [inline jplag.frontend.rust/src/test/java/de/jplag/rust/RustFrontendTest.java:77]: how about .stream().map().distinct().toList()

File: jplag.frontend.scala/src/test/java/de/jplag/scala/ScalaFrontendTest.java
Comment (dfuchss) [inline jplag.frontend.scala/src/test/java/de/jplag/scala/ScalaFrontendTest.java:103]: same as above

File: jplag.frontend.scala/src/test/java/de/jplag/scala/ScalaFrontendTest.java
Comment (dfuchss) [inline jplag.frontend.scala/src/test/java/de/jplag/scala/ScalaFrontendTest.java:163]: same as above

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/GreedyStringTiling.java:28]: We should think about IdentityHashMaps here.

File: jplag/src/main/java/de/jplag/JPlagComparison.java
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/JPlagComparison.java:11]: Maybe use `List<? extends Match>` to indicate read only list

**Review by dfuchss — CHANGES_REQUESTED** [review]
Code looks good. Some remarks to discuss,

File: jplag.frontend-utils/src/main/java/de/jplag/Language.java
Comment (JanWittler) [reply jplag.frontend-utils/src/main/java/de/jplag/Language.java:37]: Resolved by improving the documentation

File: jplag/src/main/java/de/jplag/JPlagComparison.java
Comment (JanWittler) [reply jplag/src/main/java/de/jplag/JPlagComparison.java:12]: Resolved by improving the documentation

File: jplag.frontend.scala/src/test/java/de/jplag/scala/ScalaFrontendTest.java
Comment (dfuchss) [reply jplag.frontend.scala/src/test/java/de/jplag/scala/ScalaFrontendTest.java:103]: I think this is missing :)

**Review by dfuchss — APPROVED** [review]

File: jplag.frontend.emf-metamodel/src/main/java/de/jplag/emf/util/MetamodelTreeView.java
Comment (tsaglam) [inline jplag.frontend.emf-metamodel/src/main/java/de/jplag/emf/util/MetamodelTreeView.java:45]: Probably the best solution for now, but not really a nice solution, as this method deals with two very different concerns, which makes it a but confusing for outsiders.

File: jplag.frontend.emf-metamodel/src/main/java/de/jplag/emf/util/MetamodelTreeView.java
Comment (tsaglam) [inline jplag.frontend.emf-metamodel/src/main/java/de/jplag/emf/util/MetamodelTreeView.java:51]: Should be `optionalEObject.ifPresent(eObject -> {...})`

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/GreedyStringTiling.java:35]: Now this is no longer a convenience method and thus can be removed, right?

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/GreedyStringTiling.java:37]: This change is confusing me a bit. Previously, there were two methods. One for comparison without factoring in the basecode and one for comparison where basecode matches were ignored. The second one was required for the main comparison, whereas the first one was required to calculate matching sections with the basecode for each submission in firsthand. Does it now mean after the refactoring that `compareWithBaseCode()` is the first one and `compare()` is the the second one?

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/GreedyStringTiling.java:39]: I don't really understand the sentence...

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/GreedyStringTiling.java:23]: The changes here are hard to review, we should really verify the behavior is the same. We should also test this for a run with basecode.

File: jplag/src/main/java/de/jplag/JPlagComparison.java
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/JPlagComparison.java:22]: I do not like the use of an unmodifiable list here! Upon using this record, there is no indication that the list of matches is unmodifiable. If misused, a crash will only occur at runtime. This is a badly designed API, in my opinion.

File: jplag/src/main/java/de/jplag/JPlagComparison.java
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/JPlagComparison.java:162]: Whats the reason for the removal of these methods?

File: jplag/src/main/java/de/jplag/Submission.java
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/Submission.java:155]: As already mentioned, I am not a fan of an unmodifiable list. I would prefer to return a read-only view.

File: jplag/src/main/java/de/jplag/Submission.java
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/Submission.java:149]: This is not part of your PR; but getSimilarityDivisor() should be package-private. There is no reason for it to be part of the public API.

**Review by tsaglam — CHANGES_REQUESTED** [review]
First review, no tests sanity checks done so far.

File: jplag.frontend-utils/src/main/java/de/jplag/Language.java
Comment (dfuchss) [reply jplag.frontend-utils/src/main/java/de/jplag/Language.java:35]: I mostly assume that I can only read from lists/collections from returns. Otherwise, it would get access to internals.

File: jplag/src/main/java/de/jplag/JPlagComparison.java
Comment (dfuchss) [reply jplag/src/main/java/de/jplag/JPlagComparison.java:22]: You could use List<? extends Match> but that does not guarantee that methods like remove(Object o) will be invoked.
Java has no real alternative. Always copying isn't the best option in my opinion and records should act as some kind of value type :)

File: jplag.frontend-utils/src/main/java/de/jplag/Language.java
Comment (JanWittler) [reply jplag.frontend-utils/src/main/java/de/jplag/Language.java:35]: All these things regarding unmodifiable were introduced due to [a previous review](https://github.com/jplag/JPlag/pull/534#discussion_r961579657). I don't have a strong opinion on that (except that I don't like that `List<? extends Token>` workaround) so we can also undo that.

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (JanWittler) [reply jplag/src/main/java/de/jplag/GreedyStringTiling.java:37]: I would rather say that the previous API was confusing 😄.
In the old state, the `isBaseCodeMarking` parameter was used to identify that a base code comparison is taking place and thus the token hashes are computed without considering the base code (which is only needed as the base code submission itself is tagged as base code and this has to be ignored) and then the matches are used to identify what was a base code. Otherwise, the hashes for subsequences containing a baseCode marked token were set to `-1` to avoid matches. So there was actually never the intention to "ignore" the base code markings but rather only bad separation of concerns which integrated marking of base code tokens into the core algorithm.

With the new approach, the base code submission tokens are never tagged as base code such that their subsequence hashes are always computed correctly, and the base code markings are created outside of the core algorithm.

Besides that, the `compare` and `compareWithBasecode` have the same semantics as the ones in the old version. But I will add documentation to the methods.

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (JanWittler) [reply jplag/src/main/java/de/jplag/GreedyStringTiling.java:39]:
```suggestion
        // Remove the lookup table for the current submission to trigger a regeneration as hashes will change due to the new baseCodeMarking.
```

Better like that? What I want to say is that we remove the computed hash lookup table for the submission we just computed the base code markings for, in order to trigger a regeneration of that table which considers the base code markings. As then said, this is done to increase performance, as subsequences with baseCode-marked tokens receive the `NO_VALUE` hash to ignore them during selection of possibly matching subsequences. Obviously, before computing the base code markings, the hash lookup table can't consider base code tokens and thus the hash lookup table computed during the comparison with the base code itself is not optimal.

File: jplag/src/main/java/de/jplag/JPlagComparison.java
Comment (JanWittler) [reply jplag/src/main/java/de/jplag/JPlagComparison.java:22]: So what is your suggestion? 😅

File: jplag/src/main/java/de/jplag/JPlagComparison.java
Comment (JanWittler) [reply jplag/src/main/java/de/jplag/JPlagComparison.java:162]: 1) They are not needed
2) The values regarding base code can be accessed using the `Submission#getBaseCodeComparison#similarity`

File: jplag.frontend-utils/src/main/java/de/jplag/Token.java
Comment (JanWittler) [reply jplag.frontend-utils/src/main/java/de/jplag/Token.java:10]: Good point. I expected it to work as there are tests with a FILE_END token but it seems the tests are not correctly implemented. So currently it is not working, I will fix it.

File: jplag.frontend.emf-metamodel/src/main/java/de/jplag/emf/util/MetamodelTreeView.java
Comment (JanWittler) [reply jplag.frontend.emf-metamodel/src/main/java/de/jplag/emf/util/MetamodelTreeView.java:51]: If I do that, I have to use `AtomicInteger` to be able to assign values in the lambda expression to the outside element which seemed counter-intuitive to me.

File: jplag.frontend-utils/src/main/java/de/jplag/Language.java
Comment (dfuchss) [reply jplag.frontend-utils/src/main/java/de/jplag/Language.java:35]: i guess we should simply use a shallow-copy then

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (tsaglam) [reply jplag/src/main/java/de/jplag/GreedyStringTiling.java]: Ah okay, now I get it. That makes sense! +1 for the documentation.

File: jplag/src/main/java/de/jplag/GreedyStringTiling.java
Comment (tsaglam) [reply jplag/src/main/java/de/jplag/GreedyStringTiling.java]: Ah, now I get it! Way better, thanks! 😄

File: jplag/src/main/java/de/jplag/JPlagComparison.java
Comment (tsaglam) [reply jplag/src/main/java/de/jplag/JPlagComparison.java:22]: Either we override the getter, or we turn it into a class again. Either way, I am in favor of avoiding `Collections.unmodifiableList` as much as possible, especially if it concerns the public API. However, I am heavily biased here by my pet peeve regarding the Java devs disregard for Liskov! We can also discuss our policy regarding that internally today.

File: jplag/src/main/java/de/jplag/JPlagComparison.java
Comment (tsaglam) [reply jplag/src/main/java/de/jplag/JPlagComparison.java:162]: Sounds good to me. If the Artemis crowd complains (as these methods are part of the public interface), we can always adjust.

File: jplag.frontend-utils/src/main/java/de/jplag/Token.java
Comment (JanWittler) [reply jplag.frontend-utils/src/main/java/de/jplag/Token.java:10]: Fixed in https://github.com/jplag/JPlag/pull/534/commits/01e3c582da8918f0f8b72d7d977c8d214affac0f.

I was not able to make it work in such a nice functional way as before without introducing an additional data structure (for the tuple `(line, token)`) which I found a bit over engineered. If there is one, feel free to suggest it 😄 .

File: jplag.frontend-utils/src/main/java/de/jplag/Language.java
Comment (JanWittler) [reply jplag.frontend-utils/src/main/java/de/jplag/Language.java:35]: As this is an interface which can be implemented by anyone, we cannot control what is returned and thus cannot force a shallow copy here. So for this case I would simply suggest to remove the `unmodifiable` word in the documentation again. Fine like that?

File: jplag.frontend.emf-metamodel/src/main/java/de/jplag/emf/util/MetamodelTreeView.java
Comment (JanWittler) [reply jplag.frontend.emf-metamodel/src/main/java/de/jplag/emf/util/MetamodelTreeView.java:45]: True, but this "two concerns" behaviour was already present before the refactoring. In the old state, the token was modified directly as a side effect instead of creating a new instance. So I would argue that the changes don't make things worse than they already were 😅 .

File: jplag/src/main/java/de/jplag/JPlagComparison.java
Comment (JanWittler) [reply jplag/src/main/java/de/jplag/JPlagComparison.java:22]: Agreed in maintainer meeting to use unmodifiable lists with an according documentation. Thus this resolves without further action.

File: jplag/src/main/java/de/jplag/Submission.java
Comment (JanWittler) [reply jplag/src/main/java/de/jplag/Submission.java:155]: Agreed in maintainer meeting to use unmodifiable lists with an according documentation. Thus this resolves without further action.

**Review by dfuchss — APPROVED** [review]

File: jplag.frontend-utils/src/main/java/de/jplag/Language.java
Comment (JanWittler) [reply jplag.frontend-utils/src/main/java/de/jplag/Language.java]: Documentation adapted in https://github.com/jplag/JPlag/pull/534/commits/566b736026a745fdd7473779f36bac8382450191

**Pull request merged**
